sklearn/model_selection/split.py

```diff
@@ -0,0 +1,1251 @@
+
+        Parameters
+        ----------
+        X : (Ignored, exists for compatibility.)
+        y : (Ignored, exists for compatibility.)
+        labels : (Ignored, exists for compatibility.)
+        """
+        return self.n_iter
+
+
+class PredefinedSplit(BaseCrossValidator):
+    """Predefined split cross-validator
+
+    Splits the data into training/test set folds according to a predefined
+    scheme. Each sample can be assigned to at most one test set fold, as
+    specified by the user through the ``test_fold`` parameter.
+
+    Read more in the :ref:`User Guide <cross_validation>`.
+
+    Examples
+    --------
+    >>> from sklearn.model_selection import PredefinedSplit
+    >>> X = np.array([[1, 2], [3, 4], [1, 2], [3, 4]])
+    >>> y = np.array([0, 0, 1, 1])
+    >>> test_fold = [0, 1, -1, 1]
+    >>> ps = PredefinedSplit(test_fold)
+    >>> ps.n_splits()
+    2
+    >>> print(ps)       # doctest: +NORMALIZE_WHITESPACE +ELLIPSIS
+    PredefinedSplit(test_fold=array([ 0,  1, -1,  1]))
+    >>> for train_index, test_index in ps.split():
+    ...    print("TRAIN:", train_index, "TEST:", test_index)
+    ...    X_train, X_test = X[train_index], X[test_index]
+    ...    y_train, y_test = y[train_index], y[test_index]
+    TRAIN: [1 2 3] TEST: [0]
+    TRAIN: [0 2] TEST: [1 3]
+    """
+
+    def __init__(self, test_fold):
+        self.test_fold = np.array(test_fold, dtype=np.int)
+        self.test_fold = column_or_1d(self.test_fold)
+        self.unique_folds = np.unique(self.test_fold)
+        self.unique_folds = self.unique_folds[self.unique_folds != -1]
+
+    def split(self, X=None, y=None, labels=None):
+        """Generate train/test indices to split data in train/test sets.
+
+        Parameters
+        ----------
+        X : (Ignored, exists for compatibility.)
+        y : (Ignored, exists for compatibility.)
+        labels : (Ignored, exists for compatibility.)
+        """
+        ind = np.arange(len(self.test_fold))
+        for test_index in self._iter_test_masks():
+            train_index = ind[np.logical_not(test_index)]
+            test_index = ind[test_index]
+            yield train_index, test_index
+
+    def _iter_test_masks(self):
+        """Generates boolean masks corresponding to test sets."""
+        for f in self.unique_folds:
+            test_index = np.where(self.test_fold == f)[0]
+            test_mask = np.zeros(len(self.test_fold), dtype=np.bool)
+            test_mask[test_index] = True
+            yield test_mask
+
+    def n_splits(self, X=None, y=None, labels=None):
+        """Returns the number of splitting iterations in the cross-validator
+
+        Parameters
+        ----------
+        X : (Ignored, exists for compatibility.)
+        y : (Ignored, exists for compatibility.)
+        labels : (Ignored, exists for compatibility.)
+        """
+        return len(self.unique_folds)
+
+
+class CVIterableWrapper(BaseCrossValidator):
```

Comment:
There wasn't much of a discussion on this I think... (https://github.com/scikit-learn/scikit-learn/pull/4294#issuecomment-115034149) 

Also making it private would mean users won't be able to wrap their custom cv method... but it would promote the use of new style rather than wrapping...

I am not very sure... :)